### PR TITLE
[v2] Have logging configured by a config file

### PR DESCRIPTION
**Pull request by BinaryFissionGames**
## Description of Changes
* Switch Stanza logging configuration to be done through a configuration file. (this configuration was previously a series of command line switches)

## **Please check that the PR fulfills these requirements**
- [x] Tests for the changes have been added (for bug fixes / features)
- [x] Docs have been added / updated (for bug fixes / features)
~~Add a changelog entry (for non-trivial bug fixes / features)~~ (Will be done when v2 is ready)
- [ ] CI passes

### CODE REVIEW

**Comment by BinaryFissionGames** [comment]
Alright, here's the problem with that:

The config is set using a file path. We pass that file path back to the service builder (WithConfigFile). All good so far.

Then we call build. This creates a `LogAgentBuilder`. This code here crosses the barrier into [opentelemetry-log-collection](https://pkg.go.dev/github.com/open-telemetry/opentelemetry-log-collection@v0.24.0/agent), where the interface only takes in the config file path.

Since opentelemetry-log-collection handles both the reading from disk AND the unmarshalling, there's no spot for us to intercept and modify the config into the format it should be (e.g., dropping the logging key).

opentelemetry-log-collection ALSO uses `yaml.UnmarshalStrict` (see https://github.com/open-telemetry/opentelemetry-log-collection/blob/89d49a3897633119bfe20628ff10f945d9741636/agent/config.go#L40 ), which means it errors if there are extra keys in the config, so somehow removing any keys it's not aware of (like an extra "logging" key) is necessary.

Basically, we need to make some change to opentelemetry-log-collection (which we'll have to justify, somehow), or do something like this. I'd certainly prefer if we could improve opentelemetry-log-collection to support this case.

**Comment by BinaryFissionGames** [comment]
@cpheps I decided to go ahead and just load the plugins up front. I think it solves more problems than it causes, even if it's a little odd to just ignore the option on the opentelemetry-log-collection agent builder. 

Let me know what you think about these changes.

File: service/builder.go [hunk 1]
```diff
@@ -63,33 +50,13 @@ func (b *AgentServiceBuilder) Build(ctx context.Context) (service.Service, conte
...
 func (b *AgentServiceBuilder) buildAgent() (*agent.LogAgent, error) {
 	agentBuilder := agent.NewBuilder(b.logger)
 
-	if b.configFile != nil {
-		agentBuilder = agentBuilder.WithConfigFiles([]string{*b.configFile})
-	}
-
-	if b.pluginDir != nil {
-		agentBuilder = agentBuilder.WithPluginDir(*b.pluginDir)
+	if b.config != nil {
```
Comment (cpheps) [inline service/builder.go:56]: What happens if config is nil? Can we successfully build?

Comment (BinaryFissionGames) [reply service/builder.go:56]: The agent builder will throw an error and fail to build. We could return early here, not sure what we prefer.

File: cmd/stanza/root.go [hunk 2]
```diff
@@ -87,16 +75,20 @@ func NewRootCmd() *cobra.Command {
...
 func runRoot(command *cobra.Command, _ []string, flags *RootFlags) {
-	logger := newLogger(*flags).Sugar()
+	conf, err := service.LoadConfig(flags.PluginDir, flags.ConfigFile)
+	if err != nil {
+		log.Fatalf("Failed to load config: %s", err)
+	}
+
+	logger := service.NewLogger(*conf.Logging).Sugar()
 	defer func() {
 		_ = logger.Sync()
```
[the remark this answers is not in the document]
Comment (BinaryFissionGames) [reply cmd/stanza/root.go:85]: Ok, I think I see why this was done. `make lint` complains on this line if we don't explicitly ignore the error like this.

I notice now that there's actually a lot of linting errors from the stuff we've done recently; I guess we don't have any CI check for it, which seems bizarre.

We should follow up on those linter errors in a follow up PR (probably after CI stuff is cemented). In this specific case, we might want to just leave it as is, since it's done purposefully.

Alternative is to have one of those ignore comments.

File: service/builder.go (hunk 1 above)
Comment (BinaryFissionGames) [reply service/builder.go:56]: ^ this is specifically the underlying LogAgentBuilder that will throw the error, not our service builder, if that wasn't clear.

Comment (BinaryFissionGames) [reply service/builder.go]: I think moving the error up and making it more obvious makes sense, so I've done that.

File: service/builder.go [hunk 3]
```diff
@@ -53,12 +54,14 @@ func (b *AgentServiceBuilder) Build(ctx context.Context) (service.Service, conte
 func (b *AgentServiceBuilder) buildAgent() (*agent.LogAgent, error) {
 	agentBuilder := agent.NewBuilder(b.logger)
 
-	if b.config != nil {
-		agentBuilder = agentBuilder.WithConfig(&agent.Config{
-			Pipeline: b.config.Pipeline,
-		})
+	if b.config == nil {
+		return nil, errors.New("config cannot be nil")
```
Comment (cpheps) [inline service/builder.go:58]: Nit: if this is a customer visible error we probably should say `cannot be nil`. That may not mean anything to the person setting it up.

Thinking about this though would this basically means you didn't set a config file which means you probably failed way before this? If that's true leave as is. Only someone using this as a library would hit this error in that case.

File: cmd/stanza/root.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (cpheps) [reply cmd/stanza/root.go:85]: Ok we can skip linter stuff. I'll create a ticket to setup [revive](https://github.com/mgechev/revive), the linter we're switching to, and fix all lint errors that generates.

File: service/builder.go (hunk 3 above)
Comment (BinaryFissionGames) [reply service/builder.go:58]: Yeah, this is a scenario that a customer shouldn't see. If you don't provide a config, you get this error when it can't find the default: "Failed to load config: could not read config file: open /Users/brandon/stanza/config.yaml: no such file or directory"

Basically, if this happens, it's programmer error.
